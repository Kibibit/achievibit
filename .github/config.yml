# Configuration for new-issue-welcome - https://github.com/behaviorbot/new-issue-welcome

# Comment to be posted to on first time issues
newIssueWelcomeComment: >
  Thanks for opening your first issue here! Be sure to follow the issue template!

  ![welcome](https://cdn.dribbble.com/users/27231/screenshots/2432051/welcome.gif)

# Configuration for new-pr-welcome - https://github.com/behaviorbot/new-pr-welcome

# Comment to be posted to on PRs from first time contributors in your repository
newPRWelcomeComment: >
  Thanks for opening this pull request! Please check out our contributing guidelines.

  ![welcome](https://cdn.dribbble.com/users/27231/screenshots/2432051/welcome.gif)

# Configuration for first-pr-merge - https://github.com/behaviorbot/first-pr-merge

# Comment to be posted to on pull requests merged by a first time user
firstPRMergeComment: >
  Congrats on merging your first pull request! We are so proud of you!

  ![congratz](https://media.giphy.com/media/l2Sqir5ZxfoS27EvS/giphy.gif)
  
  Hope you got a few achievements 👏 👏 👏
  
  Come again!

# It is recommend to include as many gifs and emojis as possible