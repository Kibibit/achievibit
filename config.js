module.exports = {
  'port': process.env.PORT || 3141,
  'database': 'mongodb://node:noder@novus.modulusmongo.net:27017/Iganiq8o',
  'secret': 'ilovescotchscotchyscotchscotch'
};
