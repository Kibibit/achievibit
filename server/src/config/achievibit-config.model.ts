import { Exclude, Expose } from 'class-transformer';
import { defaultMetadataStorage } from 'class-transformer/cjs/storage';
import {
  IsBoolean,
  IsIn,
  IsNumber,
  IsOptional,
  IsString,
  IsUrl,
  Matches,
  Validate
} from 'class-validator';
import { validationMetadatasToSchemas } from 'class-validator-jsonschema';
import { v1 as uuidv1 } from 'uuid';

import { JsonSchema } from './json-schema.validator';

export const NODE_ENVIRONMENT_OPTIONS = [
  'development',
  'production',
  'test',
  'devcontainer'
];
export const SMEE_IO_REGEX = /^https:\/\/(?:www\.)?smee\.io\/[a-zA-Z0-9_-]+\/?/;
export const ENDPONT_PATH_REGEX = /^([\w]+)?(\/[\w-]+)*$/;

@Exclude()
export class AchievibitConfig {
  @Expose()
  @IsString()
  @IsIn(NODE_ENVIRONMENT_OPTIONS)
  @Validate(JsonSchema, [
    'Tells which env file to use'
  ])
  nodeEnv = 'development';

  @Expose()
  @IsNumber()
  @Validate(JsonSchema, [
    'Set server port'
  ])
  port = 10101;

  @Expose()
  @IsOptional()
  @IsString()
  @IsUrl({ protocols: [ 'mongodb', 'mongodb+srv' ], require_tld: false }, {
    message: '$property should be a valid mongodb URL'
  })
  @Validate(JsonSchema, [
    'DB connection URL. Expects a mongodb db for connections'
  ])
  dbUrl: string;

  @Expose()
  @Matches(SMEE_IO_REGEX)
  @IsUrl()
  @Validate(JsonSchema, [
    'Used to create a custom repeatable smee webhook url instead of ',
    'Generating a random one'
  ])
  webhookProxyUrl = `https://smee.io/achievibit-${ uuidv1() }`;

  @Expose()
  @Matches(ENDPONT_PATH_REGEX)
  @Validate(JsonSchema, [
    'proxy should sent events to this url for achievibit'
  ])
  // @Transform((url) => trim(url as any, '/'))
  webhookDestinationUrl = 'events';

  @IsBoolean()
  @Validate(JsonSchema, [
    'Create a file made out of the internal config. This is mostly for ',
    'merging command line, environment, and file variables to a single instance'
  ])
  saveToFile = false;

  constructor(partial: Partial<AchievibitConfig> = {}) {
    Object.assign(this, partial);
  }

  toJsonSchema() {
    const configJsonSchema = validationMetadatasToSchemas({
      classTransformerMetadataStorage: defaultMetadataStorage,
      additionalConverters: {
        JsonSchema: (meta) => ({
            description: meta.constraints.join('')
        })
      }
    }).AchievibitConfig;

    delete configJsonSchema.properties.nodeEnv;
    configJsonSchema.required.splice(
      configJsonSchema.required.indexOf('nodeEnv'),
      1
    );

    // console.log(configJsonSchema);

    return configJsonSchema;
  }
}
