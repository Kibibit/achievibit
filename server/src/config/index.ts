/**
 * @file Automatically generated by barrelsby.
 */

export * from './achievibit-config.model';
export * from './config.module';
export * from './config.service';
export * from './json-schema.validator';
export * from './winston.config';
