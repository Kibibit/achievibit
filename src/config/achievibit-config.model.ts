import { Exclude, Expose, Transform } from 'class-transformer';
import { IsBoolean, IsIn, IsNumber, IsOptional, IsString, IsUrl, Matches } from 'class-validator';
import { trim } from 'lodash';

import { dtoMockGenerator } from '../dto.mock-generator';

export const NODE_ENVIRONMENT_OPTIONS = [ 'development', 'production', 'test' ];
export const SMEE_IO_REGEX = /^https:\/\/(?:www\.)?smee\.io\/[a-zA-Z0-9_-]+\/?/;
export const ENDPONT_PATH_REGEX = /^([\w]+)?(\/[\w-]+)*$/;

@Exclude()
export class AchievibitConfig {
  @Expose()
  @IsString()
  @IsIn(NODE_ENVIRONMENT_OPTIONS)
  nodeEnv = 'development';

  @Expose()
  @IsNumber()
  port = 10101;

  @Expose()
  @IsOptional()
  @IsString()
  @IsUrl({ protocols: [ 'mongodb', 'mongodb+srv' ] }, {
    message: '$property should be a valid mongodb URL'
  })
  dbUrl: string;

  @Expose()
  @Matches(SMEE_IO_REGEX)
  @IsUrl()
  webhookProxyUrl = `https://smee.io/achievibit-${ dtoMockGenerator.guid() }`;

  @Expose()
  @Matches(ENDPONT_PATH_REGEX, null, {
    message: 'should be a server endpoint path matching the regex: $constraint1'
  })
  @Transform((url) => trim(url, '/'))
  webhookDestinationUrl = 'events';

  @Expose()
  @IsBoolean()
  saveToFile = false;

  constructor(partial: Partial<AchievibitConfig> = {}) {
    Object.assign(this, partial);
  }
}
