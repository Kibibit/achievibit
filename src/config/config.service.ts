import { Logger } from '@nestjs/common';
import { classToPlain, Exclude } from 'class-transformer';
import { validateSync } from 'class-validator';
import findRoot from 'find-root';
import { writeJson } from 'fs-extra';
import { camelCase, get } from 'lodash';
import nconf from 'nconf';
import { join } from 'path';
import SmeeClient from 'smee-client';

import { ConfigValidationError } from '@kb-errors';

import { AchievibitConfig } from './achievibit-config.model';

const appRoot = findRoot(__dirname);
const environment = get(process, 'env.NODE_ENV', 'development');
const eventLogger: Logger = new Logger('SmeeEvents');
(eventLogger as any).info = eventLogger.log;
const configFilePath = join(appRoot, `${ environment }.env.json`);

nconf
  .argv({
    parseValues: true
  })
  .env({
    lowerCase: true,
    parseValues: true,
    transform: transformToLowerCase
  })
  .file({ file: configFilePath });

let smee: SmeeClient;
let events: any;
let configService: ConfigService;

/**
 * This is a **Forced Singleton**.
 * This means that even if you try to create
 * another ConfigService, you'll always get the
 * first one.
 */
@Exclude()
export class ConfigService extends AchievibitConfig {
  private readonly logger: Logger = new Logger('ConfigService');

  private readonly mode: string = environment;

  get smee(): SmeeClient {
    return smee;
  }

  get events(): any {
    return events;
  }

  constructor(passedConfig?: Partial<AchievibitConfig>) {
    super();

    if (!passedConfig && configService) { return configService; }

    const config = passedConfig || nconf.get();
    const envConfig = this.validateInput(config);

    // attach configuration to this service
    Object.assign(this, envConfig);

    if (this.mode === 'development' || passedConfig.nodeEnv === 'development') {
      if (!smee) {
        smee = new SmeeClient({
          source: this.webhookProxyUrl,
          target: `http://localhost:${ this.port }/${ this.webhookDestinationUrl }`,
          logger: eventLogger
        });
      }

      if (!events) {
        this.logger.log('Starting to listen to events from Proxy');
        events = this.smee.start();
      }
    } else {
      this.closeEvents();

      smee = undefined;
      events = undefined;
    }

    if (this.saveToFile) {
      writeJson(configFilePath, classToPlain(this), { spaces: 2 });
    }

    configService = this;
  }

  closeEvents() {
    return this.events && this.events.close();
  }

  toPlainObject() {
    return classToPlain(this);
  }

  /**
   * Ensures all needed variables are set, and returns the validated JavaScript object
   * including the applied default values.
   */
  private validateInput(envConfig: Record<string, any>): Partial<AchievibitConfig> {
    const achievibitConfig = new AchievibitConfig(envConfig);
    const validationErrors = validateSync(achievibitConfig);

    if (validationErrors.length > 0) {
      throw new ConfigValidationError(validationErrors);
    }
    return classToPlain(achievibitConfig);
  }
}

function transformToLowerCase(obj: { key: string; value: string }) {
  const camelCasedKey = camelCase(obj.key);

  obj.key = camelCasedKey;

  return camelCasedKey && obj;
}
