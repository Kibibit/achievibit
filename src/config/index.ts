/**
 * @file Automatically generated by barrelsby.
 */

export * from './config.module';
export * from './config.service';
