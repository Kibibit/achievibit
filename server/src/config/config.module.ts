import { Module } from '@nestjs/common';

import { ConfigService } from './config.service';

@Module({
  providers: [
    {
      provide: ConfigService,
      useValue: new ConfigService()
    }
  ],
  exports: [ ConfigService ]
})
export class ConfigModule { }
