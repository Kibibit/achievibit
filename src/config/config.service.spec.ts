import { chain, times } from 'lodash';

import { DtoMockGenerator } from '@kb-dev-tools';
import { ConfigValidationError } from '@kb-errors';

import { AchievibitConfig, NODE_ENVIRONMENT_OPTIONS, SMEE_IO_REGEX } from './achievibit-config.model';
import { ConfigService } from './config.service';

// not a jest component so testing it as a node module
describe('ConfigService', () => {

  describe('Forced Singleton and baypass', () => {
    let configService: ConfigService;
    beforeEach(() => {
      configService = new ConfigService({
        webhookProxyUrl: 'https://smee.io/achievibit-test'
      });
    });

    afterEach(async (done) => {
      configService.closeEvents();

      done();
    });

    it('should be defined', async (done) => {
      expect(configService).toBeDefined();

      done();
    });

    it('should return the same instance if initiated without an input', async (done) => {
      expect(new ConfigService()).toBe(configService);

      done();
    });

    it('should create a new instance when passed an override', async (done) => {
      expect(new ConfigService({} as AchievibitConfig)).not.toBe(configService);

      done();
    });

    it('should set default values to everything that needs one', async (done) => {
      expect(configService.toPlainObject()).toMatchSnapshot();

      done();
    });
  });

  describe('Smee & Events', () => {
    it('should NOT initial smee and events on production', async (done) => {
      const productionService = new ConfigService({
        nodeEnv: 'production'
      });

      expect(productionService.smee).toBeUndefined();
      expect(productionService.events).toBeUndefined();

      productionService.closeEvents();

      done();
    });

    it('should initial smee and events on development', async (done) => {
      const productionService = new ConfigService({
        nodeEnv: 'development'
      });

      expect(productionService.smee).toBeDefined();
      expect(productionService.events).toBeDefined();

      productionService.closeEvents();

      done();
    });
  });

  describe('Saving configuration', () => {
    it.todo('should allow user to save configuration to file');
  });

  describe('Validations', () => {
    describe('nodeEnv', () => {

      chain(NODE_ENVIRONMENT_OPTIONS)
        .forEach((nodeEnv: string) => {
          it(`should ACCEPT ${ nodeEnv }`, async (done) => {
            const serviceWrapper = () => new ConfigService({ nodeEnv });

            expect(serviceWrapper().toPlainObject).toBeDefined();

            serviceWrapper().closeEvents();

            done();
          });
        })
        .value();

      it('should REJECT other values', async (done) => {

        const nodeEnv = 'value_not_allowed';

        const wrongEnv = () => new ConfigService({ nodeEnv });

        const nodeEnvNumber = 4;

        const wrongEnvType = () => new (ConfigService as any)({ nodeEnv: nodeEnvNumber });

        expect(wrongEnv).toThrowError(ConfigValidationError);
        expect(wrongEnv).toThrowErrorMatchingSnapshot();

        expect(wrongEnvType).toThrowError(ConfigValidationError);
        expect(wrongEnvType).toThrowErrorMatchingSnapshot();

        done();
      });

    });

    describe('port', () => {
      it('should ACCEPT numbers', async (done) => {
        const port = DtoMockGenerator.integer();
        const service = new ConfigService({ port });

        expect(service.port).toBe(port);

        done();
      });

      it('should REJECT values other than numbers', async (done) => {
        const stringPort = () => new (ConfigService as any)({ port: 'hello' });
        const ObjectPort = () => new (ConfigService as any)({ port: {} });

        expect(stringPort).toThrowError(ConfigValidationError);
        expect(stringPort).toThrowErrorMatchingSnapshot();

        expect(ObjectPort).toThrowError(ConfigValidationError);
        expect(ObjectPort).toThrowErrorMatchingSnapshot();

        done();
      });

    });

    describe('dbUrl', () => {
      it('should ACCEPT empty value', async (done) => {
        const service = new ConfigService({ dbUrl: undefined });

        expect(service.dbUrl).toBeUndefined();

        service.closeEvents();

        done();
      });

      it('should ACCEPT localhost mongodb URL', async (done) => {
        const localMongodbUrl = 'mongodb://localhost:27017';

        const configService = new ConfigService({ dbUrl: localMongodbUrl });

        expect(configService.dbUrl).toBe(localMongodbUrl);

        done();
      });

      it('should ACCEPT valid mongodb URLS', async (done) => {
        const mongodbProtocolUrls = times(10, () => DtoMockGenerator.mongodbUrl());

        chain(mongodbProtocolUrls)
          .keyBy()
          .mapValues((mongodbUrl) => new ConfigService({ dbUrl: mongodbUrl }))
          .forEach((configService, mongodbUrl) => {
            expect(configService.dbUrl).toBe(mongodbUrl);
          })
          .value();

        done();
      });

      it('should REJECT non-mongodb URLS', async (done) => {
        const invalidUrl = () => new ConfigService({ dbUrl: 'https://google.com/' });

        expect(invalidUrl).toThrowError(ConfigValidationError);
        expect(invalidUrl).toThrowErrorMatchingSnapshot();

        done();
      });
    });

    describe('webhookProxyUrl', () => {
      it('should ACCEPT smee.io URLS', async (done) => {
        const smeeProtocolUrls = times(10, () => DtoMockGenerator.randexp(SMEE_IO_REGEX).gen());

        console.log(smeeProtocolUrls);

        chain(smeeProtocolUrls)
          .keyBy()
          .mapValues((smeeProtocolUrl) => new ConfigService({ webhookProxyUrl: smeeProtocolUrl }))
          .forEach((configService, smeeProtocolUrl) => {
            configService.closeEvents();
            expect(configService.webhookProxyUrl).toBe(smeeProtocolUrl);
          })
          .value();

        done();
      });

      it('should REJECT non-smee URLS', async (done) => {
        const invalidUrl = () => new ConfigService({ webhookProxyUrl: 'https://google.com/' });

        expect(invalidUrl).toThrowError(ConfigValidationError);
        expect(invalidUrl).toThrowErrorMatchingSnapshot();

        done();
      });

      it('should REJECT non-URLS', async (done) => {
        const invalidUrl = () => new ConfigService({ webhookProxyUrl: 'hello world' });

        expect(invalidUrl).toThrowError(ConfigValidationError);
        expect(invalidUrl).toThrowErrorMatchingSnapshot();

        done();
      });
    });
  });
});
