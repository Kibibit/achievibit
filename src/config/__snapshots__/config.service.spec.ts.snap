// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigService Forced Singleton and baypass should set default values to everything that needs one 1`] = `
Object {
  "dbUrl": undefined,
  "nodeEnv": "development",
  "port": 10101,
  "saveToFile": false,
  "webhookDestinationUrl": "events",
  "webhookProxyUrl": "https://smee.io/achievibit-test",
}
`;

exports[`ConfigService Validations dbUrl should REJECT non-mongodb URLS 1`] = `
"
============================
 property: dbUrl
 value: https://google.com/
============================
 - dbUrl should be a valid mongodb URL

"
`;

exports[`ConfigService Validations nodeEnv should REJECT other values 1`] = `
"
==========================
 property: nodeEnv
 value: value_not_allowed
==========================
 - nodeEnv must be one of the following values: development,production,test

"
`;

exports[`ConfigService Validations nodeEnv should REJECT other values 2`] = `
"
===================
 property: nodeEnv
 value: 4
===================
 - nodeEnv must be one of the following values: development,production,test
 - nodeEnv must be a string

"
`;

exports[`ConfigService Validations port should REJECT values other than numbers 1`] = `
"
================
 property: port
 value: hello
================
 - port must be a number conforming to the specified constraints

"
`;

exports[`ConfigService Validations port should REJECT values other than numbers 2`] = `
"
========================
 property: port
 value: [object Object]
========================
 - port must be a number conforming to the specified constraints

"
`;

exports[`ConfigService Validations webhookProxyUrl should REJECT non-URLS 1`] = `
"
===========================
 property: webhookProxyUrl
 value: hello world
===========================
 - webhookProxyUrl must be an URL address
 - webhookProxyUrl must match /^https:\\\\/\\\\/(?:www\\\\.)?smee\\\\.io\\\\/[a-zA-Z0-9_-]+\\\\/?/ regular expression

"
`;

exports[`ConfigService Validations webhookProxyUrl should REJECT non-smee URLS 1`] = `
"
============================
 property: webhookProxyUrl
 value: https://google.com/
============================
 - webhookProxyUrl must match /^https:\\\\/\\\\/(?:www\\\\.)?smee\\\\.io\\\\/[a-zA-Z0-9_-]+\\\\/?/ regular expression

"
`;
